### PR TITLE
lmdb: print available disk space after each db resize

### PR DESCRIPTION
**Pull request by sumogr**
and warn us hard if it is below 4 Gb

### CODE REVIEW

File: src/blockchain_db/lmdb/db_lmdb.cpp [hunk 1]
```diff
@@ -518,10 +518,10 @@ void BlockchainLMDB::do_resize(uint64_t increase_size)
   const uint64_t add_size = 1LL << 30;
 
   // check disk capacity
+  boost::filesystem::path path(m_folder);
+  boost::filesystem::space_info si = boost::filesystem::space(path);
```
Comment (selsta) [inline src/blockchain_db/lmdb/db_lmdb.cpp:522]: What about the try/catch block?

File: src/blockchain_db/lmdb/db_lmdb.cpp [hunk 2]
```diff
@@ -574,7 +574,10 @@ void BlockchainLMDB::do_resize(uint64_t increase_size)
   if (result)
     throw0(DB_ERROR(lmdb_error("Failed to set new mapsize: ", result).c_str()));
 
-  MGINFO("LMDB Mapsize increased." << "  Old: " << mei.me_mapsize / (1024 * 1024) << "MiB" << ", New: " << new_mapsize / (1024 * 1024) << "MiB");
+  if ((si.available / (1024 * 1024)) > 2000)
+  MGINFO("LMDB Mapsize increased." << "  Old: " << mei.me_mapsize / (1024 * 1024) << "MiB" << ", New: " << new_mapsize / (1024 * 1024) << "MiB" << ", Available disk remaining: " << si.available / (1024 * 1024) << "MiB" );
```
Comment (selsta) [inline src/blockchain_db/lmdb/db_lmdb.cpp:578]: Indentation seems to be missing

Comment (sumogr) [reply src/blockchain_db/lmdb/db_lmdb.cpp]: done

File: src/blockchain_db/lmdb/db_lmdb.cpp (hunk 1 above)
Comment (sumogr) [reply src/blockchain_db/lmdb/db_lmdb.cpp]: there `si` redeclared

Comment (selsta) [reply src/blockchain_db/lmdb/db_lmdb.cpp]: boost::filesystem::space can still fail

Comment (sumogr) [reply src/blockchain_db/lmdb/db_lmdb.cpp]: oh got it now

Comment (sumogr) [reply src/blockchain_db/lmdb/db_lmdb.cpp]: How about this boost's magic potion @selsta ? (i think it can be used instead of that try/catch block as well)

Comment (selsta) [reply src/blockchain_db/lmdb/db_lmdb.cpp]: This does not look correct to me. si.available is uintmax_t and not an optional type. This does not catch filesystem_error exceptions.

Comment (sumogr) [reply src/blockchain_db/lmdb/db_lmdb.cpp]: ok you most likely misunderstand the use of boost::optional 's class template.
_DESCRIPTION:
The class template boost::optional manages an optional contained value (a value that or may not be present).A common use of boost::optional is in cases when the return value of a function may fail in some cases._

Comment (selsta) [reply src/blockchain_db/lmdb/db_lmdb.cpp]: I know the use case of boost optional but this requires a boost::optional return value which is not the case here. See here as an example:

https://www.boost.org/doc/libs/1_66_0/libs/optional/doc/html/boost_optional/quick_start.html

But I will leave this to the more experienced devs :P

Comment (sumogr) [reply src/blockchain_db/lmdb/db_lmdb.cpp]: Actually boost::option maybe used for any nullable object see this example https://www.boost.org/doc/libs/1_65_1/libs/optional/doc/html/index.html
I am closing this and leave more experienced devs :P to keep on suggesting code to your project

Comment (selsta) [reply src/blockchain_db/lmdb/db_lmdb.cpp]: Your example also has an boost::optional return value: `boost::optional<int> getConfigParam`

> I am closing this and leave more experienced devs :P to keep on suggesting code to your project

Not sure why you closed this now. I only said that I will leave this review to devs who are more experienced than me.

Comment (sumogr) [reply src/blockchain_db/lmdb/db_lmdb.cpp]: Pardon me for being jumpy.

File: src/blockchain_db/lmdb/db_lmdb.cpp [hunk 3]
```diff
@@ -573,8 +574,22 @@ void BlockchainLMDB::do_resize(uint64_t increase_size)
...
     throw0(DB_ERROR(lmdb_error("Failed to set new mapsize: ", result).c_str()));
+  
+  boost::filesystem::path path(m_folder);
+  boost::filesystem::space_info si = boost::filesystem::space(path);
+  boost::optional<uint64_t> space_available = si.available;
 
-  MGINFO("LMDB Mapsize increased." << "  Old: " << mei.me_mapsize / (1024 * 1024) << "MiB" << ", New: " << new_mapsize / (1024 * 1024) << "MiB");
+  if (space_available)
+  {
+   if ((*space_available / (1024 * 1024)) < 4294)
```
Comment (moneromooo-monero) [inline src/blockchain_db/lmdb/db_lmdb.cpp:584]: Why this particular odd threshold, out of curiosity ?

File: src/blockchain_db/lmdb/db_lmdb.cpp [hunk 4]
```diff
@@ -573,8 +574,22 @@ void BlockchainLMDB::do_resize(uint64_t increase_size)
...
+     MGINFO_RED("Warning you are running low on disk space!\n" << "LMDB Mapsize increased." << "  Old: " << mei.me_mapsize / (1024 * 1024) << "MiB" << ", New: " << new_mapsize / (1024 * 1024) << "MiB" << ", Available disk remaining: " << *space_available / (1024 * 1024) << "MiB" );
+   else if ((*space_available / (1024 * 1024)) >= 4294)
```
Comment (moneromooo-monero) [inline src/blockchain_db/lmdb/db_lmdb.cpp:586]: FWIW, that test will always trigger, no ?

File: src/blockchain_db/lmdb/db_lmdb.cpp (hunk 1 above)
Comment (moneromooo-monero) [reply src/blockchain_db/lmdb/db_lmdb.cpp]: selsta is right. Assigning a return value will happen if no exception is thrown by the call. If an exception is thrown, that assignment will be skipped anyway. The optional is pointless in this case. Use try/catch and get rid of the optional. Using optional would only make sense for an API which was designed to use it, which this seems not to.

**Comment by dEBRUYNE-1** [comment]
@sumogr - ping. There are a few open questions.

File: src/blockchain_db/lmdb/db_lmdb.cpp (hunk 3 above)
Comment (sumogr) [reply src/blockchain_db/lmdb/db_lmdb.cpp:584]: well i empirically presume that below 4 Gb there will most likely be no more room for the next resize, especially when synced or approaching the end of the chain while syncing. If a more reasonable threshold is proposed i ll change it

File: src/blockchain_db/lmdb/db_lmdb.cpp (hunk 4 above)
Comment (sumogr) [reply src/blockchain_db/lmdb/db_lmdb.cpp:586]: yes

**Pull request closed without merging**